src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,208 @@
+        return oldVal;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> node = node(index);
+        unlink(node);
+        return node.item;
     }
 
     @Override
     public boolean remove(T t) {
+        if (t == null) {
+            for (Node<T> x = first; x != null; x = x.next) {
+                if (x.item == null) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> x = first; x != null; x = x.next) {
+                if (t.equals(x.item)) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> node(int index) {
+        checkPositionIndex(index);
+
+        if (index < (size >> 1)) {
+            Node<T> x = first;
+            for (int i = 0; i < index; i++) {
+                x = x.next;
+            }
+            return x;
+        } else {
+            Node<T> x = last;
+            for (int i = size - 1; i > index; i--) {
+                x = x.prev;
+            }
+            return x;
+        }
+    }
+
+    private void checkPositionIndex(int index) {
+        if (!isPositionIndex(index)) {
+            throw new IndexOutOfBoundsException(outOfBoundsMsg(index));
+        }
+    }
+
+    private boolean isPositionIndex(int index) {
+        return (index >= 0 && index < size);
+    }
+
+    private void unlink(Node<T> x) {
+        Node<T> next = x.next;
+        Node<T> prev = x.prev;
+
+        if (prev == null) {
+            first = next;
+        } else {
+            prev.next = next;
+        }
+
+        if (next == null) {
+            last = prev;
+        } else {
+            next.prev = prev;
+            x.next = null;
+        }
+
+        size--;
+    }
+
+    private void linkLast(T e) {
+        Node<T> l = last;
+        Node<T> newNode = new Node<T>(l, e, null);
+        last = newNode;
+        if (l == null) {
+            first = newNode;
+        } else {
+            l.next = newNode;
+        }
+        size++;
+    }
+
+    private void linkBefore(T t, Node<T> succ) {
+        Node<T> pred = succ.prev;
+        Node<T> newNode = new Node<T>(pred, t, succ);
+        succ.prev = newNode;
+        if (pred == null) {
+            first = newNode;
+        } else {
+            pred.next = newNode;
+        }
+        size++;
+    }
+
+    private static class Node<T> {
+        T item;
+        Node<T> prev;
+        Node<T> next;
+
+        Node(Node<T> prev, T element, Node<T> next) {
+            this.item = element;
+            this.prev = prev;
+            this.next = next;
+        }
+    }
+
+    private String outOfBoundsMsg(int index) {
```

Comment:
do we need this method? we use it only one time, so let's remove it

Reply:
@floriangeyer21 what about this?

Reply:
This method is redundant! You are using this only in one place. This is method pollution. Remove this method